swarm/src/connection.rs

```diff
@@ -138,38 +149,38 @@ where
             match self.handler.poll(cx)? {
                 Poll::Pending => {}
                 Poll::Ready(handler_wrapper::Event::OutboundSubstreamRequest(user_data)) => {
-                    self.muxing.open_substream(user_data);
-                    continue;
+                    self.open_info.push_back(user_data);
+                    continue; // Poll handler until exhausted.
                 }
                 Poll::Ready(handler_wrapper::Event::Custom(event)) => {
                     return Poll::Ready(Ok(Event::Handler(event)));
                 }
             }
 
-            // Perform I/O on the connection through the muxer, informing the handler
-            // of new substreams.
-            match self.muxing.poll(cx)? {
-                Poll::Pending => {}
-                Poll::Ready(SubstreamEvent::InboundSubstream { substream }) => {
-                    self.handler
-                        .inject_substream(substream, SubstreamEndpoint::Listener);
-                    continue;
-                }
-                Poll::Ready(SubstreamEvent::OutboundSubstream {
-                    user_data,
-                    substream,
-                }) => {
+            if !self.open_info.is_empty() {
+                if let Poll::Ready(substream) = self.muxing.poll_outbound(cx)? {
+                    let user_data = self
+                        .open_info
+                        .pop_front()
+                        .expect("`open_info` is not empty");
                     let endpoint = SubstreamEndpoint::Dialer(user_data);
                     self.handler.inject_substream(substream, endpoint);
-                    continue;
-                }
-                Poll::Ready(SubstreamEvent::AddressChange(address)) => {
-                    self.handler.inject_address_change(&address);
-                    return Poll::Ready(Ok(Event::AddressChange(address)));
+                    continue; // Go back to the top, handler can potentially make progress again.
```

Comment:
Side note, I like this style of:

- When one made progress
- but one does not return
- start from the top via `continue` as other sub components might be able to make progress.

Reply:
I do too although it is a pattern I had to get used too and it is not trivial to understand when you get started with async Rust.